backend/src/main/java/com/bakdata/conquery/apiv1/query/concept/specific/external/CQExternal.java

```diff
@@ -78,42 +92,87 @@ public class CQExternal extends CQElement {
 	 * @implNote FK: I would prefer to implement this as a guava table, but they cannot be deserialized with Jackson so we implement the Table manually.
 	 */
 	@InternalOnly
-	@Getter
+	@Getter(AccessLevel.PRIVATE)
 	private Map<Integer, Map<String, List<String>>> extra;
 
-	public CQExternal(@NotEmpty List<String> format, @NotEmpty String[][] values) {
+	public CQExternal(List<String> format, @NotEmpty String[][] values, boolean onlySingles) {
 		this.format = format;
 		this.values = values;
+		this.onlySingles = onlySingles;
 	}
 
+	@JsonIgnore
+	@ValidationMethod(message = "Headers are not unique")
+	@SuppressWarnings("unused")
+	public boolean isHeadersUnique() {
+		try {
+			// Try to create a Set. Fails with IllegalArgumentException if duplicates exists.
+			// Ignore return value.
+			Set.of(values[0]);
+			return true;
+		}
+		catch (IllegalArgumentException e) {
+			log.warn("Headers are not unique.", e);
+		}
+		return false;
+	}
 
 	@Override
 	public QPNode createQueryPlan(QueryPlanContext context, ConceptQueryPlan plan) {
 		if (valuesResolved == null) {
 			throw new IllegalStateException("CQExternal needs to be resolved before creating a plan");
 		}
 
-		// Allocate at once, the maximum possible size
-		final Map<String, ConstantValueAggregator<List<String>>> extraAggregators = new HashMap<>(format.size());
+		final String[] extraHeaders = Streams.zip(
+													 Arrays.stream(headers),
+													 format.stream(),
+													 (header, format) -> format.equals(FORMAT_EXTRA) ? header : null
+											 )
+											 .filter(Objects::nonNull)
+											 .toArray(String[]::new);
 
-		if (extra != null) {
-			for (int col = 0; col < format.size(); col++) {
-				if (!format.get(col).equals(FORMAT_EXTRA)) {
-					continue;
-				}
+		if (!onlySingles) {
```

Comment:
Die 2 Branches sind gar nicht mehr so klein, kannst du den getrennten Teil bitte in Methoden auslagern?

Reply:
sind jetzt einzelne methoden